File: packages/coterie-py/src/coterie/nodes/consensus_engine.py
"""Consensus engine node for `consensus` mode.

N participant agents independently report findings. The engine clusters
semantically equivalent findings (via an LLM) and labels each cluster by
agreement ratio:

- `confirmed`        — agreement_ratio >= confirm_threshold
- `needs-verification` — agreement_count >= 2 but below threshold
- `unverified`       — single agent only
"""

import json
from typing import Any

from coterie.core.llm.base import LLMClient
from coterie.core.state import CoterieState
from coterie.nodes.auditor import parse_findings_from_output

ENGINE_SYSTEM = """You cluster code-review findings into groups of semantically equivalent items.
Input: a JSON array of findings, each with an `agent_id` field identifying which agent produced it.
Output: a JSON array of clusters. Each cluster:
{
  "description": "<canonical phrasing>",
  "category": "<most common category>",
  "severity": "<highest severity in cluster>",
  "supporting_agents": ["<agent_id>", ...],
  "member_indices": [<int>, ...]
}

Two findings belong in the same cluster when they identify the same defect, even if
worded differently. Be strict — do not merge unrelated findings just because they share a category.
Return only the JSON array, no prose, no markdown.
"""


def make_consensus_engine_node(llm: LLMClient | None = None):
    def engine(state: CoterieState) -> dict[str, Any]:
        cfg = state["config"]
        cons_cfg = cfg.get("consensus") or {}
        engine_cfg = cons_cfg.get("engine") or {}
        threshold = engine_cfg.get("confirm_threshold", 0.5)
        participants = cons_cfg.get("participants") or [a["id"] for a in cfg["agents"]]
        n_participants = len(participants)

        flat: list[dict] = []
        for run in state.get("runs") or []:
            if run.get("role") != "consensus-participant" or run["agent_id"] not in participants:
                continue
            for f in parse_findings_from_output(run.get("stdout", "")):
                if isinstance(f, dict) and "description" in f:
                    flat.append({**f, "agent_id": run["agent_id"]})

        if not flat:
            return {
                "mode_state": {**(state.get("mode_state") or {}), "consensus_findings": []},
                "status": "done",
            }

        if llm is None:
            # No engine LLM — fall back to one-cluster-per-finding (no merging).
            clusters = [
                {
                    "description": f["description"],
                    "category": f.get("category", "unknown"),
                    "severity": f.get("severity", "low"),
                    "supporting_agents": [f["agent_id"]],
                    "member_indices": [i],
                }
                for i, f in enumerate(flat)
            ]
        else:
            raw = llm.chat(ENGINE_SYSTEM, [{"role": "user", "content": f"Cluster these findings:\n{json.dumps(flat)}"}])
            try:
                clusters = json.loads(raw.strip().lstrip("`").rstrip("`"))
                if not isinstance(clusters, list):
                    raise ValueError("not a list")
            except (json.JSONDecodeError, ValueError):
                clusters = [
                    {
                        "description": f["description"],
                        "category": f.get("category", "unknown"),
                        "severity": f.get("severity", "low"),
                        "supporting_agents": [f["agent_id"]],
                        "member_indices": [i],
                    }
                    for i, f in enumerate(flat)
                ]

        consensus: list[dict] = []
        for c in clusters:
            supporters = list(set(c.get("supporting_agents", [])))
            ratio = len(supporters) / max(n_participants, 1)
            # A single supporter is never `confirmed`, regardless of threshold.
            # Multiple supporters can be `confirmed` (above threshold) or
            # `needs-verification` (below).
            if len(supporters) <= 1:
                label = "unverified"
            elif ratio >= threshold:
                label = "confirmed"
            else:
                label = "needs-verification"
            consensus.append(
                {
                    "description": c.get("description", ""),
                    "category": c.get("category", "unknown"),
                    "severity": c.get("severity", "low"),
                    "agreement_count": len(supporters),
                    "agreement_ratio": ratio,
                    "label": label,
                    "supporting_agents": supporters,
                }
            )

        return {
            "mode_state": {**(state.get("mode_state") or {}), "consensus_findings": consensus},
            "status": "done",
        }

    return engine

File: packages/coterie-js/src/nodes/consensusEngine.ts
import { parseJsonLoose } from "../core/json.js";
import type { LLMClient } from "../core/llm/base.js";
import type { CoterieState } from "../core/state.js";
import { parseFindings } from "./auditor.js";

const ENGINE_SYSTEM = `You cluster code-review findings into groups of semantically equivalent items.
Input: a JSON array of findings, each with an \`agent_id\` field identifying which agent produced it.
Output: a JSON array of clusters. Each cluster:
{
  "description": "<canonical phrasing>",
  "category": "<most common category>",
  "severity": "<highest severity in cluster>",
  "supporting_agents": ["<agent_id>", ...],
  "member_indices": [<int>, ...]
}

Two findings belong in the same cluster when they identify the same defect, even if
worded differently. Be strict — do not merge unrelated findings just because they share a category.
Return only the JSON array, no prose, no markdown.`;

export function makeConsensusEngineNode(llm: LLMClient | null) {
  return async (state: CoterieState) => {
    const cfg = state.config;
    const consCfg = cfg.consensus ?? {};
    const engineCfg = consCfg.engine ?? {};
    // `consensus.threshold` is the documented key; `engine.confirm_threshold` is
    // an older alias. Honor either (threshold wins) before falling back.
    const threshold = consCfg.threshold ?? engineCfg.confirm_threshold ?? 0.5;
    const participants = (consCfg.participants ?? (cfg.agents as any[]).map((a: any) => a.id)) as string[];
    const n = participants.length;

    const flat: any[] = [];
    for (const run of state.runs ?? []) {
      if (run.role !== "consensus-participant" || !participants.includes(run.agent_id)) continue;
      for (const f of parseFindings(run.stdout)) {
        if (f && typeof f === "object" && typeof f.description === "string") {
          flat.push({ ...f, agent_id: run.agent_id });
        }
      }
    }

    if (flat.length === 0) {
      return {
        mode_state: { ...(state.mode_state ?? {}), consensus_findings: [] },
        status: "done" as const,
      };
    }

    let clusters: any[];
    if (!llm) {
      clusters = flat.map((f, i) => ({
        description: f.description,
        category: f.category ?? "unknown",
        severity: f.severity ?? "low",
        supporting_agents: [f.agent_id],
        member_indices: [i],
      }));
    } else {
      const raw = await llm.chat(ENGINE_SYSTEM, [
        { role: "user", content: `Cluster these findings:\n${JSON.stringify(flat)}` },
      ]);
      const parsed = parseJsonLoose(raw);
      if (Array.isArray(parsed)) {
        clusters = parsed;
      } else {
        clusters = flat.map((f, i) => ({
          description: f.description,
          category: f.category ?? "unknown",
          severity: f.severity ?? "low",
          supporting_agents: [f.agent_id],
          member_indices: [i],
        }));
      }
    }

    const consensus = clusters.map((c) => {
      // Intersect with the real participant set: a hallucinated/duplicate agent id
      // must not inflate agreement (ratio could otherwise exceed 1.0).
      const supporters = [...new Set<string>(c.supporting_agents ?? [])].filter((s) =>
        participants.includes(s),
      );
      const ratio = supporters.length / Math.max(n, 1);
      const label =
        supporters.length <= 1 ? "unverified" : ratio >= threshold ? "confirmed" : "needs-verification";
      return {
        description: c.description ?? "",
        category: c.category ?? "unknown",
        severity: c.severity ?? "low",
        agreement_count: supporters.length,
        agreement_ratio: ratio,
        participant_count: n,
        label,
        supporting_agents: supporters,
      };
    });

    return {
      mode_state: { ...(state.mode_state ?? {}), consensus_findings: consensus },
      status: "done" as const,
    };
  };
}
